test/unit/math/rev/fun/as_array_or_scalar_test.cpp

```diff
@@ -0,0 +1,28 @@
+#include <stan/math/rev.hpp>
+#include <test/unit/util.hpp>
+
+#include <gtest/gtest.h>
+
+TEST(MathFunRev, as_array_or_scalar_var_value_matrix) {
+  int n = 100;
+  const Eigen::MatrixXd a_val = Eigen::MatrixXd::Random(n, n);
+  stan::math::var_value<Eigen::MatrixXd> a(a_val);
+  const auto& tmp = stan::math::as_array_or_scalar(a);
```

Comment:
Yeah add some sorta type check here. These tests would pass if we just returned the input matrix.